File: 동적계획법/동전교환.py
"""
import sys
# sys.stdin = open("input.txt", 'r')

if __name__ == "__main__":
    n = int(input())
    coin = list(map(int, input().split()))
    m = int(input())
    memo = [0] * (m + 1)

    for i in range(n):
        for j in range(coin[i], m + 1):
            if i == 0:
                memo[j] = memo[j - coin[0]] + 1
            else:
                memo[j] = min(memo[j], memo[j - coin[i]] + 1)

    print(memo[-1])
"""
import sys
# sys.stdin = open("input.txt", 'r')

if __name__ == "__main__":
    n = int(input())
    coin = list(map(int, input().split()))
    m = int(input())
    memo = [500] * (m + 1)
    memo[0] = 0

    for i in range(n):
        for j in range(coin[i], m + 1):
            memo[j] = min(memo[j], memo[j - coin[i]] + 1)
    print(memo[-1])

File: 동적계획법/동전교환.cpp
#include <bits/stdc++.h>
using namespace std;

int main(void){
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	// freopen("input.txt", "r", stdin);
	
	int n;
	int i, j, payback;
	
	cin >> n;
	vector<int> coin(n);	
	for(i = 0; i < n; i++){
		cin >> coin[i];
	}
	
	cin >> payback;
	vector<int> memo(payback + 1, 500);
	memo[0] = 0;
	for(i = 0; i < n; i++){
		for(j = coin[i]; j <= payback; j++){
			memo[j] = min(memo[j], memo[j - coin[i]] + 1);
		}
	}
	cout << memo[payback];
	
	return 0;
}
